contracts/cw20-ics20/src/ibc.rs

```diff
@@ -0,0 +1,248 @@
+    Ok(())
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+/// record the channel in CHANNEL_INFO
+pub fn ibc_channel_connect(
+    deps: DepsMut,
+    _env: Env,
+    channel: IbcChannel,
+) -> Result<IbcBasicResponse, ContractError> {
+    // we need to check the counter party version in try and ack (sometimes here)
+    enforce_order_and_version(&channel)?;
+
+    let info = ChannelInfo {
+        id: channel.endpoint.channel_id,
+        counterparty_endpoint: channel.counterparty_endpoint,
+        connection_id: channel.connection_id,
+    };
+    CHANNEL_INFO.save(deps.storage, &info.id, &info)?;
+
+    // TODO: add events/attributes here?
+    let res = IbcBasicResponse::default();
+    Ok(res)
+}
+
+fn enforce_order_and_version(channel: &IbcChannel) -> Result<(), ContractError> {
+    if channel.version != ICS20_VERSION {
+        return Err(ContractError::InvalidIbcVersion {
+            version: channel.version.clone(),
+        });
+    }
+    if let Some(version) = &channel.counterparty_version {
+        if version != ICS20_VERSION {
+            return Err(ContractError::InvalidIbcVersion {
+                version: version.clone(),
+            });
+        }
+    }
+    if channel.order != ICS20_ORDERING {
+        return Err(ContractError::OnlyOrderedChannel {});
+    }
+    Ok(())
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn ibc_channel_close(
+    _deps: DepsMut,
+    _env: Env,
+    _channel: IbcChannel,
+) -> Result<IbcBasicResponse, ContractError> {
+    // TODO: what to do here?
+    // we will have locked funds that need to be returned somehow
+    unimplemented!();
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+/// Check to see if we have any balance here
+/// We should not return an error if possible, but rather an acknowledgement of failure
+pub fn ibc_packet_receive(
+    deps: DepsMut,
+    _env: Env,
+    packet: IbcPacket,
+) -> Result<IbcReceiveResponse, ContractError> {
+    // TODO: don't let error leak
+    let msg: Ics20Packet = from_binary(&packet.data)?;
+    let channel = packet.src.channel_id;
+    let denom = msg.denom;
+    let amount = Uint128::from(msg.amount);
+    CHANNEL_STATE.update(
+        deps.storage,
+        (&channel, &denom),
+        |orig| -> Result<_, ContractError> {
+            // this will return error if we don't have the funds there to cover the request (or no denom registered)
+            let mut cur = orig.ok_or(ContractError::InsufficientFunds {})?;
+            cur.outstanding = (cur.outstanding - amount)?;
+            Ok(cur)
+        },
+    )?;
+
+    // if we have funds, now send the tokens to the requested recipient
+    let to_send = Amount::from_parts(denom, amount);
+    let msg = send_amount(to_send, HumanAddr::from(msg.receiver))?;
+    let res = IbcReceiveResponse {
+        acknowledgement: ack_success()?,
+        messages: vec![msg],
+        // TODO: similar event messages like ibctransfer module
```

Comment:
Following event is emitted:
```go
			types.EventTypePacket,
			sdk.NewAttribute(sdk.AttributeKeyModule, types.ModuleName),
			sdk.NewAttribute(types.AttributeKeyReceiver, data.Receiver),
			sdk.NewAttribute(types.AttributeKeyDenom, data.Denom),
			sdk.NewAttribute(types.AttributeKeyAmount, fmt.Sprintf("%d", data.Amount)),
			sdk.NewAttribute(types.AttributeKeyAckSuccess, fmt.Sprintf("%t", err != nil)),
```

Reply:
ohh.... this is a very important and very major issue!

packet and acknowledgements are never stored in the iavl tree, but rather their hashes are there and you need to get the preimage from the events.

When we submit `SendPacket`, this is handled in wasmd and *should* emit such a packet (need to check). Since the contract cannot create custom event types, we need to also auto-emit such an event in wasmd for acknowledgements. Otherwise we cannot relay such packets.

Reply:
Okay, investigated a bit more. The ChannelKeeper manages all the essential events.

On [`SendPacket`](https://github.com/cosmos/cosmos-sdk/blob/v0.42.0/x/ibc/core/04-channel/keeper/packet.go#L120-L135) and [`WriteAcknowledgement`](https://github.com/cosmos/cosmos-sdk/blob/v0.42.0/x/ibc/core/04-channel/keeper/packet.go#L367-L382)

Reply:
Update: We return the acknowledgement from `IbcModule.OnRecvPacket` in `x/wasm`.

That is the [same pattern that is used in `ibctransfer`](https://github.com/cosmos/cosmos-sdk/blob/v0.42.0/x/ibc/applications/transfer/module.go#L322-L354) and when digging deeper, I see that the acknowledgement is [written by the ibc msg server](https://github.com/cosmos/cosmos-sdk/blob/v0.42.0/x/ibc/core/keeper/msg_server.go#L445-L458).

So, it looks like we don't need to do anything here, but add some app-specific attributes